substrate/frame/remote-proxy/src/lib.rs

```diff
@@ -0,0 +1,498 @@
+			BlockNumberFor<T>,
+		>>::RemoteBlockNumber;
+	type RemoteAccountIdOf<T, I> = <<T as Config<I>>::RemoteProxy as RemoteProxyInterface<
+		<T as frame_system::Config>::AccountId,
+		<T as pallet_proxy::Config>::ProxyType,
+		BlockNumberFor<T>,
+	>>::RemoteAccountId;
+	type RemoteHasherOf<T, I> = <<T as Config<I>>::RemoteProxy as RemoteProxyInterface<
+		<T as frame_system::Config>::AccountId,
+		<T as pallet_proxy::Config>::ProxyType,
+		BlockNumberFor<T>,
+	>>::RemoteHasher;
+	type RemoteHashOf<T, I> = <<T as Config<I>>::RemoteProxy as RemoteProxyInterface<
+		<T as frame_system::Config>::AccountId,
+		<T as pallet_proxy::Config>::ProxyType,
+		BlockNumberFor<T>,
+	>>::RemoteHash;
+	type RemoteProxyTypeOf<T, I> = <<T as Config<I>>::RemoteProxy as RemoteProxyInterface<
+		<T as frame_system::Config>::AccountId,
+		<T as pallet_proxy::Config>::ProxyType,
+		BlockNumberFor<T>,
+	>>::RemoteProxyType;
+	type WeightInfoOf<T, I> = <T as Config<I>>::WeightInfo;
+
+	#[pallet::pallet]
+	pub struct Pallet<T, I = ()>(_);
+
+	/// Stores the last [`Config::MaxStorageRootsToKeep`] block to storage root mappings of the
+	/// target chain.
+	#[pallet::storage]
+	pub type BlockToRoot<T: Config<I>, I: 'static = ()> = StorageValue<
+		_,
+		BoundedVec<(RemoteBlockNumberOf<T, I>, RemoteHashOf<T, I>), T::MaxStorageRootsToKeep>,
+		ValueQuery,
+	>;
+
+	/// Configuration trait.
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config + pallet_proxy::Config {
+		/// The maximum number of storage roots to keep.
+		///
+		/// The storage roots are used to validate the remote proofs. The more we keep in storage,
+		/// the older the proof can be. This is not only seen as a maximum number, but also as the
+		/// maximum difference between the latest and the oldest storage root stored. This means
+		/// that if the chain for example did not progress for `MaxStorageRootsToKeep` blocks, only
+		/// the latest added storage root will be available for validating proofs.
+		type MaxStorageRootsToKeep: Get<u32>;
+
+		/// The interface for interacting with the remote proxy.
+		type RemoteProxy: RemoteProxyInterface<
+			Self::AccountId,
+			Self::ProxyType,
+			BlockNumberFor<Self>,
+		>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	impl<T: Config<I>, I: 'static> OnSystemEvent for Pallet<T, I> {
+		fn on_validation_data(validation_data: &PersistedValidationData) {
+			let Some((block, hash)) = T::RemoteProxy::block_to_storage_root(validation_data) else {
+				return;
+			};
+
+			// Update the block to root mappings.
+			BlockToRoot::<T, I>::mutate(|roots| {
+				let delete_up_to =
+					block.clone().saturating_sub(T::MaxStorageRootsToKeep::get().into());
+
+				while roots.first().is_some_and(|f| f.0 <= delete_up_to) {
+					roots.remove(0);
+				}
+
+				// We always remove all the old items before, thus there should always be space in
+				// the vector.
+				let _res = roots.try_push((block, hash));
```

Comment:
Please check if the last item block number is already in there. This can happen because we sometimes build multiple parachain blocks on the same relay chain block.

Please also add a test for it.